src/ucp/core/ucp_am.c

```diff
@@ -1079,42 +1112,51 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_recv_data_nbx,
     ucp_dt_recv_state_init(&req->recv.state, buffer, datatype, count);
     req->recv.length   = ucp_dt_length(datatype, count, buffer,
                                        &req->recv.state);
-    req->recv.mem_type = ucp_request_get_memory_type(context, buffer,
-                                                     req->recv.length, param);
-    req->recv.am.desc  = (ucp_recv_desc_t*)rts - 1;
+    req->recv.mem_type = mem_type;
+    req->recv.am.desc  = desc;
 
     ucp_request_set_callback_param(param, recv_am, req, recv.am);
 
-    ucs_assertv(req->recv.length >= rts->super.size,
-                "rx buffer too small %zu, need %zu",
-                req->recv.length, rts->super.size);
+    if ((desc->flags & UCP_RECV_DESC_FLAG_RNDV) && (count > 0)) {
```

Comment:
didn't we already check count==0?

Reply:
yes, but that was for  no immediate completion case (UCP_OP_ATTR_FLAG_NO_IMM_CMPL))

Reply:
can we somehow unite the code?

Reply:
ok

Reply:
seems no good  way to unite. The only common thing is  calling `ucp_am_rndv_send_ats(worker, data_desc, UCS_OK);`. Others are different (plz note one place uses `ucp_dt_unpack_only` other  `ucp_request_recv_data_unpack`)